File: lecture_notes/10 refactoring/6-hash.py
import json

def parse(point):
    if type(point) != dict:
        point = json.loads(point)
    else:
        pass
    return point


def move(offset, point):
    for key in point:
        if key in offset:
            point[key] = point[key] + offset[key]
    path = {**point, **offset}
    return path
    

polyline = [
    { "x": 0, "y": 0 },
    { "x": 10, "y": 10 },
    '{ "x": 20, "y": 20 }',
    { "x": 30, "y": 30 }]

start_point = { "x": 10, "y": -5 }
parsed = list(map(parse, polyline))
path = list(map(move, start_point, parsed))
print(path)

File: lecture_notes/10 refactoring/6-hash.js
'use strict';

const shift =
  ({ a, b }) =>
    ({ x, y }) => ({ x: x + a, y: y + b });

const typeCheck = { string: JSON.parse, object: (x) => x };

const genCheck = (line) => typeCheck[typeof line](line);

const polyline = [
  { x: 0, y: 0 },
  { x: 10, y: 10 },
  '{ "x": 20, "y": 20 }',
  { x: 30, y: 30 },
];

const offset = shift({ a: 10, b: -5 });
const checked = polyline.map(genCheck);
const path = checked.map(offset);
console.log({ path });
